Terminal.Gui/Views/Scroll/Scroll.cs

```diff
@@ -0,0 +1,214 @@
+        {
+            Position = Math.Max (Position - barSize, 0);
+        }
+        else if (mouseEvent.Flags.HasFlag (MouseFlags.Button1Pressed) && location > sliderPos.bottomRight)
+        {
+            Position = Math.Min (Position + barSize, Size - barSize);
+        }
+        else if ((mouseEvent.Flags == MouseFlags.WheeledDown && Orientation == Orientation.Vertical)
+                 || (mouseEvent.Flags == MouseFlags.WheeledRight && Orientation == Orientation.Horizontal))
+        {
+            Position = Math.Min (Position + 1, Size - barSize);
+        }
+        else if ((mouseEvent.Flags == MouseFlags.WheeledUp && Orientation == Orientation.Vertical)
+                 || (mouseEvent.Flags == MouseFlags.WheeledLeft && Orientation == Orientation.Horizontal))
+        {
+            Position = Math.Max (Position - 1, 0);
+        }
+        else if (mouseEvent.Flags == MouseFlags.Button1Clicked)
+        {
+            if (_slider.Frame.Contains (mouseEvent.Position))
+            {
+                return _slider.OnMouseEvent (mouseEvent);
+            }
+        }
+
+        return base.OnMouseEvent (mouseEvent);
+    }
+
+    // TODO: Move this into "ScrollSlider" and override it there. Scroll can then subscribe to _slider.LayoutComplete and call AdjustSlider.
+    // QUESTION: I've been meaning to add a "View.FrameChanged" event (fired from LayoutComplete only if Frame has changed). Should we do that as part of this PR?
+    // QUESTION: Note I *did* add "View.ViewportChanged" in a previous PR.
+
+    /// <summary>Virtual method called when <see cref="Position"/> has changed. Raises <see cref="PositionChanged"/>.</summary>
+    protected virtual void OnPositionChanged (int position) { PositionChanged?.Invoke (this, new (in position)); }
+
+    /// <summary>
+    ///     Virtual method called when <see cref="Position"/> is changing. Raises <see cref="PositionChanging"/>, which is
+    ///     cancelable.
+    /// </summary>
+    protected virtual CancelEventArgs<int> OnPositionChanging (int currentPos, int newPos)
+    {
+        CancelEventArgs<int> args = new (ref currentPos, ref newPos);
+        PositionChanging?.Invoke (this, args);
+
+        return args;
+    }
+
+    /// <summary>Virtual method called when <see cref="Size"/> has changed. Raises <see cref="SizeChanged"/>.</summary>
+    protected void OnSizeChanged (int size) { SizeChanged?.Invoke (this, new (in size)); }
+
+    internal void AdjustScroll ()
+    {
+        if (SupView is { })
+        {
+            X = Orientation == Orientation.Vertical ? 0 : 1;
+            Y = Orientation == Orientation.Vertical ? 1 : 0;
+            Width = Orientation == Orientation.Vertical ? Dim.Fill () : Dim.Fill (1);
+            Height = Orientation == Orientation.Vertical ? Dim.Fill (1) : Dim.Fill ();
+        }
+
+        _slider.AdjustSlider ();
+        SetScrollText ();
+    }
+
+    /// <inheritdoc/>
+    internal override void OnLayoutComplete (LayoutEventArgs args)
```

Comment:
Do you know precisely what situations require this override?

I like to ask that each time I'm tempted to ovrride OnLayoutStarted/Complete because often it means I'm not doing something else right. 

Sometimes it is needed though.

Reply:
Oh, and are you sure it `OnLayoutStarted` doesn't make more sense? in AdjustScroll you neve test any super/subviews, so it seems so.

Reply:
> Do you know precisely what situations require this override?
> 

Mainly for console resize, where recalculation, others than layout itself, may cause change on some calculation that will envolve layout changes.

> I like to ask that each time I'm tempted to ovrride OnLayoutStarted/Complete because often it means I'm not doing something else right.
> 
Sometimes using the `OnLayoutStarted` may cause recursive execution that may cause undesired behaviors.

> Sometimes it is needed though.

Yes, indeed.

> Oh, and are you sure it OnLayoutStarted doesn't make more sense? in AdjustScroll you neve test any super/subviews, so it seems so.

When the `Scroll` is used alone I want allow the user the Pos/Dim he want. When the super-view is a `ScrollBar` only makes sense let the `ScrollBar` handle the correct position and dimension. When the `ScrollBar` is used alone, allow the user also set the Pos/Dim he want.